// @angularclass/form-validators/dist/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var forms_1 = require('@angular/forms');
var validators = require('./validators');
function createValidator(type) {
    var validator = validators[type];
    return function customValidator(c) {
        if (c.value !== undefined && c.value !== null && validator(c.value)) {
            return null;
        }
        return (_a = {}, _a[type] = true, _a);
        var _a;
    };
}
var AcIsCreditCardValidator = (function () {
    function AcIsCreditCardValidator() {
    }
    AcIsCreditCardValidator = __decorate([
        core_1.Directive({
            selector: '' +
                '[ac-is-credit-card][ngModel],' +
                '[ac-is-credit-card][formControl],' +
                '[acIsCredit-card][ngModel],' +
                '[acIsCredit-card][formControl]' +
                '',
            providers: [{
                    provide: forms_1.NG_VALIDATORS,
                    useValue: createValidator('isCreditCard'),
                    multi: true
                }]
        }), 
        __metadata('design:paramtypes', [])
    ], AcIsCreditCardValidator);
    return AcIsCreditCardValidator;
}());
exports.AcIsCreditCardValidator = AcIsCreditCardValidator;
var AcIsBase64Validator = (function () {
    function AcIsBase64Validator() {
    }
    AcIsBase64Validator = __decorate([
        core_1.Directive({
            selector: '' +
                '[ac-is-base-64][ngModel],' +
                '[ac-is-base-64][formControl],' +
                '[acIsBase64][ngModel],' +
                '[acIsBase64][formControl]' +
                '',
            providers: [{
                    provide: forms_1.NG_VALIDATORS,
                    useValue: createValidator('isBase64'),
                    multi: true
                }]
        }), 
        __metadata('design:paramtypes', [])
    ], AcIsBase64Validator);
    return AcIsBase64Validator;
}());
exports.AcIsBase64Validator = AcIsBase64Validator;
var AcIsEmailValidator = (function () {
    function AcIsEmailValidator() {
    }
    AcIsEmailValidator = __decorate([
        core_1.Directive({
            selector: '' +
                '[ac-is-email][ngModel],' +
                '[ac-is-email][formControl],' +
                '[acIsEmail][ngModel],' +
                '[acIsEmail][formControl]' +
                '',
            providers: [{
                    provide: forms_1.NG_VALIDATORS,
                    useValue: createValidator('isEmail'),
                    multi: true
                }]
        }), 
        __metadata('design:paramtypes', [])
    ], AcIsEmailValidator);
    return AcIsEmailValidator;
}());
exports.AcIsEmailValidator = AcIsEmailValidator;
var AcIsJSONValidator = (function () {
    function AcIsJSONValidator() {
    }
    AcIsJSONValidator = __decorate([
        core_1.Directive({
            selector: '' +
                '[ac-is-json][ngModel],' +
                '[ac-is-json][formControl],' +
                '[acIsJson][ngModel],' +
                '[acIsJson][formControl]' +
                '',
            providers: [{
                    provide: forms_1.NG_VALIDATORS,
                    useValue: createValidator('isJSON'),
                    multi: true
                }]
        }), 
        __metadata('design:paramtypes', [])
    ], AcIsJSONValidator);
    return AcIsJSONValidator;
}());
exports.AcIsJSONValidator = AcIsJSONValidator;
var AcIsPhoneValidator = (function () {
    function AcIsPhoneValidator() {
    }
    AcIsPhoneValidator = __decorate([
        core_1.Directive({
            selector: '' +
                '[ac-is-phone][ngModel],' +
                '[ac-is-phone][formControl],' +
                '[acIsPhone][ngModel],' +
                '[acIsPhone][formControl]' +
                '',
            providers: [{
                    provide: forms_1.NG_VALIDATORS,
                    useValue: createValidator('isPhone'),
                    multi: true
                }]
        }), 
        __metadata('design:paramtypes', [])
    ], AcIsPhoneValidator);
    return AcIsPhoneValidator;
}());
exports.AcIsPhoneValidator = AcIsPhoneValidator;
var AcIsUUIDValidator = (function () {
    function AcIsUUIDValidator() {
    }
    AcIsUUIDValidator = __decorate([
        core_1.Directive({
            selector: '' +
                '[ac-is-uuid][ngModel],' +
                '[ac-is-uuid][formControl],' +
                '[acIsUUID][ngModel],' +
                '[acIsUUID][formControl]' +
                '',
            providers: [{
                    provide: forms_1.NG_VALIDATORS,
                    useValue: createValidator('isUUID'),
                    multi: true
                }]
        }), 
        __metadata('design:paramtypes', [])
    ], AcIsUUIDValidator);
    return AcIsUUIDValidator;
}());
exports.AcIsUUIDValidator = AcIsUUIDValidator;
exports.ANGULARCLASS_FORM_VALIDATOR_DIRECTIVES = [
    AcIsBase64Validator,
    AcIsCreditCardValidator,
    AcIsEmailValidator,
    AcIsJSONValidator,
    AcIsPhoneValidator,
    AcIsUUIDValidator
];
